django/thunderstore/repository/api/experimental/serializers.py

```diff
@@ -95,3 +96,132 @@ class Meta:
             "has_nsfw_content",
             "categories",
         )
+
+
+class PackageUploadAuthorNameField(serializers.SlugRelatedField):
+    def __init__(self, *args, **kwargs):
+        kwargs["slug_field"] = "name"
+        super().__init__(*args, **kwargs)
+
+    def get_queryset(self):
+        return UploaderIdentity.objects.filter(
+            members__user=self.context["request"].user,
+        )
+
+
+class PackageUploadCategoriesField(serializers.RelatedField):
+    def get_queryset(self):
+        return PackageCategory.objects.filter(
+            community=self.context["request"].community,
+        )
+
+    def to_representation(self, value):
+        return [c.pk for c in value]
+
+    def to_internal_value(self, data):
+        if not isinstance(data, list):
+            raise serializers.ValidationError("Not a list")
+
+        out = []
+        for category_pk in data:
+            category = PackageCategory.objects.filter(pk=category_pk).first()
+            if not category:
+                raise serializers.ValidationError(f"{category_pk} category not found")
+            out.append(category)
+        return out
+
+
+class PackageUploadSerializer(serializers.Serializer):
+    author_name = PackageUploadAuthorNameField()
+    categories = PackageUploadCategoriesField()
+    has_nsfw_content = serializers.BooleanField()
+    pk = serializers.IntegerField(read_only=True)
+    file = serializers.FileField(write_only=True)
+
+    def _create_form(self, data) -> PackageUploadForm:
+        request = self.context["request"]
+        return PackageUploadForm(
+            request.user,
+            data.get("author_name"),
+            request.community,
+            data={
+                "categories": data.get("categories"),
+                "has_nsfw_content": data.get("has_nsfw_content"),
+            },
+            files={"file": data.get("file")},
+        )
+
+    def validate(self, data):
+        form = self._create_form(data)
+        if not form.is_valid():
+            raise serializers.ValidationError("Invalid")
+        return data
+
+    def create(self, validated_data) -> PackageVersion:
+        form = self._create_form(validated_data)
+        form.is_valid()
+        return form.save()
+
+
+class UploaderIdentitySerializer(serializers.ModelSerializer):
+    class Meta:
+        model = UploaderIdentity
+        fields = (
+            "name",
+            "pk",
+        )
+
+    def create(self, *args, **kwargs):
+        raise Exception()
```

Comment:
There's probably a better error here. `NotImplementedError` wouldn't make sense as that implies it is going to be in the future, but these serializers are purely used for reading. There is also `read_only` but this would require setting it for every field(?).

Reply:
These could also be removed from code coverage as they should be classed as programming errors

Reply:
You could implement your own ReadOnlySerializer base class to use, probably would come in handy in the future as well